### PR TITLE
Fix(eos_cli_config_gen,eos_designs): Dont configure access group on interface when access group is defined on session level

### LINKED ISSUE #4524: Bug(eos_cli_config_gen, eos_designs): Monitor-session needs correction
### Issue Summary

Monitor-session needs correction. When we create assign an access-group at session level, we cant assign a different access-group to the source interfaces.

```
s1-leaf1(config)#monitor session m1 ip access-group ipv4ACL
! Access list ipv4ACL not configured. Assigning anyway.
s1-leaf1(config)#monitor session m1 source ethernet 1 ipv6 access-group abcd
! Source Ethernet1 in Session m1 is applied with access list ipv4ACL with type ip instead of access list abcd with type ipv6. Only one access list type per session is supported.
```

### Which component(s) of AVD impacted

eos_cli_config_gen

### How do you run AVD ?

PyAVD

### Steps to reproduce

```shell
Create a monitor session and add ip access-group command.
Try adding ipv6 access-group command with source interface.
```

### Relevant log output

_No response_

### Contributing Guide

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by laxmikantchintakindi**
## Change Summary

Dont configure access group to interface when access group is defined on session level

## Related Issue(s)

Fixes #4524

## Component(s) name

`arista.avd.eos_cli_config_gen`
`arista.avd.eos_designs`

## Proposed changes
Dont configure access group to interface when access group is defined on session level

```
s1-leaf1(config)#monitor session m1 ip access-group ipv4ACL
! Access list ipv4ACL not configured. Assigning anyway.
s1-leaf1(config)#monitor session m1 source ethernet 1 ipv6 access-group abcd
! Source Ethernet1 in Session m1 is applied with access list ipv4ACL with type ip instead of access list abcd with type ipv6. Only one access list type per session is supported.
```
### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [ ] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/DC1-SVC3A.cfg [hunk 1]
```diff
@@ -874,9 +874,9 @@ ip prefix-list PL-MLAG-PEER-VRFS
    seq 10 permit 10.255.251.6/31
 !
 monitor session MonitoringSessionServer18WithDest mac access-group mac_acl
-monitor session MonitoringSessionServer18WithDest source Ethernet25 rx ip access-group MyIpACL priority 5
-monitor session MonitoringSessionServer18WithDest source Ethernet28 tx mac access-group MyMacACL priority 5
-monitor session MonitoringSessionServer18WithDest source Port-Channel27 tx mac access-group MyMacACL priority 5
+monitor session MonitoringSessionServer18WithDest source Ethernet25 rx
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/DC1-SVC3A.cfg:892]: Please add another session so we still have tests of mac access groups and both ip and mac with priority.

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-sessions.j2 [hunk 2]
```diff
@@ -8,17 +8,20 @@
...
 {%         for source in monitor_session.sources | arista.avd.natural_sort('name') %}
 {%             set source_cli = "monitor session " ~ monitor_session.name ~ " source " ~ source.name %}
 {%             if source.direction is arista.avd.defined %}
 {%                 set source_cli = source_cli ~ " " ~ source.direction %}
 {%             endif %}
-{%             if source.access_group.type is arista.avd.defined and source.access_group.name is arista.avd.defined %}
-{%                 set source_cli = source_cli ~ " " ~ source.access_group.type ~ " access-group " ~ source.access_group.name %}
-{%                 if source.access_group.priority is arista.avd.defined %}
-{%                     set source_cli = source_cli ~ " priority " ~ source.access_group.priority %}
+{%             if session_ag | arista.avd.default(false) is arista.avd.defined(false) %}
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-sessions.j2]: This could also be combined with the inner `if` instead of an extra nested level.
```suggestion
{%             if session_ag is not arista.avd.defined %}
```

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/DC1-SVC3A.cfg (hunk 1 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/DC1-SVC3A.cfg:892]: Also update the test data with comments saying that these options are ignored because of the ACL on the session.

**Review by ClausHolbechArista — CHANGES_REQUESTED** [review]

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-sessions.j2 (hunk 2 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-sessions.j2]: Fixed.

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/DC1-SVC3A.cfg (hunk 1 above)
Comment (laxmikantchintakindi) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/DC1-SVC3A.cfg:892]: Added.

**Comment by laxmikantchintakindi** [comment]
Can I update  eos_designs for monitor_sesion with this logic?

**Comment by ClausHolbechArista** [comment]
> Can I update eos_designs for monitor_sesion with this logic?

Sure, then update the scope on PR title to be both roles.

**Comment by laxmikantchintakindi** [comment]
> > Can I update eos_designs for monitor_sesion with this logic?
> 
> Sure, then update the scope on PR title to be both roles.

Fixed.

File: python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py [hunk 3]
```diff
@@ -46,12 +46,13 @@ def monitor_sessions(self: AvdStructuredConfigConnectedEndpoints) -> list | None
                     "name": session["interface"],
                     "direction": get(session, "source_settings.direction"),
                 }
-                if (access_group := get(session, "source_settings.access_group")) is not None:
-                    source["access_group"] = {
-                        "type": access_group.get("type"),
-                        "name": access_group.get("name"),
-                        "priority": access_group.get("priority"),
-                    }
+                if get(merged_settings, "session_settings.access_group") is None:
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py]: In eos_designs we should not just ignore stuff. Instead we should raise an error so the user is made aware.

File: python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py [hunk 4]
```diff
@@ -46,7 +47,12 @@ def monitor_sessions(self: AvdStructuredConfigConnectedEndpoints) -> list | None
                     "name": session["interface"],
                     "direction": get(session, "source_settings.direction"),
                 }
-                if (access_group := get(session, "source_settings.access_group")) is not None:
+
+                if get(merged_settings, "session_settings.access_group") is not None:
+                    if get(session, "source_settings.access_group") is not None:
+                        msg = f"For {self.shared_utils.hostname}, an ACL is set on {session['interface']} even though ACL is set for monitor session {session_name}."
```
Comment (MaheshGSLAB) [inline python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py]: please add a negative test

Comment (MaheshGSLAB) [reply python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py]: Also fix the pre-commit

Comment (MaheshGSLAB) [reply python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py]: Can we give similar error message 

"For {self.shared_utils.hostname}, source `{session['interface']}` in Session `{session_name}` was applied with access list `<acl_name>`. Please remove this access list first before applying another access list.

**Review by MaheshGSLAB — APPROVED** [review]

File: python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py (hunk 3 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py]: Added

File: python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py [hunk 5]
```diff
@@ -46,7 +47,15 @@ def monitor_sessions(self: AvdStructuredConfigConnectedEndpoints) -> list | None
                     "name": session["interface"],
                     "direction": get(session, "source_settings.direction"),
                 }
-                if (access_group := get(session, "source_settings.access_group")) is not None:
+
+                if get(merged_settings, "session_settings.access_group") is not None:
+                    if get(session, "source_settings.access_group") is not None:
+                        msg = (
+                            f"For {self.shared_utils.hostname}, an ACL is set on {session['interface']}"
+                            f" even though ACL is set for monitor session {session_name} in connected endpoints settings."
```
Comment (Vibhu-gslab) [inline python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py]:
```suggestion
                            f"Can not set ACL for {self.shared_utils.hostname} on interface {session['interface']} since it is already present on monitor session {session_name} in connected endpoints settings."
```

File: ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/host_vars/connected-endpoints-monitor-session-mismatch-acl.yml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+                type: ip
+                name: acl2
+            session_settings:
+              access_group:
+                type: ip
+                name: acl1
+
+expected_error_message: >-
+  For connected-endpoints-monitor-session-mismatch-acl,
+  an ACL is set on Port-Channel14 even though ACL is set for monitor session DMF in connected endpoints settings.
```
Comment (Vibhu-gslab) [inline ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/host_vars/connected-endpoints-monitor-session-mismatch-acl.yml]:
```suggestion
  Can not set ACL for connected-endpoints-monitor-session-mismatch-acl on interface Port-Channel14 since it is already present on monitor session DMF in connected endpoints settings.
```

File: python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py (hunk 5 above)
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py]: I would reword this to be a bit more guiding with a better context

```
Cannot set an ACL both under the session {session_name} session_settings source_settings for adapter <PATH TO ADAPTER>
```

Comment (gmuloc) [reply python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py]: to get the context you can inject it in _monitor_session_configs. You will need to build different contexts for connected_endpoints and netwrok_ports

File: python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py [hunk 7]
```diff
@@ -7,6 +7,7 @@
 from functools import cached_property
 from typing import TYPE_CHECKING
 
+from pyavd._errors import AristaAvdError
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py]: Use AristaAvdInvalidInputsError

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-sessions.j2 [hunk 8]
```diff
@@ -8,17 +8,20 @@
 !
 {%     for monitor_session in monitor_sessions | arista.avd.natural_sort('name') %}
 {%         if monitor_session.access_group.type is arista.avd.defined and monitor_session.access_group.name is arista.avd.defined %}
+{%             set session_ag = true %}
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-sessions.j2]: use `session_access_group` for the name

File: python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py (hunk 5 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py]: Fixed.

Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py]: Fixed.

File: python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py (hunk 7 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_designs/structured_config/connected_endpoints/monitor_sessions.py]: Done.

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-sessions.j2 (hunk 8 above)
Comment (laxmikantchintakindi) [reply python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/monitor-sessions.j2]: Done.

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-sessions.md [hunk 9]
```diff
@@ -94,6 +94,24 @@ interface Management1
 | -------- | ------ |
 | Destinations | - |
 
+##### myMonitoringSession4
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-sessions.md:97]: In line 89 the per source ACL is still printed even though it is ignored in the config.

**Review by ClausHolbechArista — CHANGES_REQUESTED** [review]

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-sessions.md (hunk 9 above)
Comment (laxmikantchintakindi) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-sessions.md:97]: Shall I print access_group set at session level instead of empty cell?

Comment (alexeygorbunov) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-sessions.md:97]: > In line 89 the per source ACL is still printed even though it is ignored in the config.

Isn't line 89 OK? It covers myMonitoringSession3 which has no ACL-based collisions. Is issue more on line 71 (with myMonitoringSession2)?

> Shall I print access_group set at session level instead of empty cell?

It's technically already covered under `Destinations and Session Settings` of the monitor session, so would lead to duplication

**Comment by alexeygorbunov** [comment]
Would it be useful to have a 3rd negative test case where part of the config is coming from `connected_endpoints` and other part is coming from `network_ports`?

```yaml
loopback_ipv4_pool: 192.168.1.0/24

type: l2leaf
l2leaf:
  defaults:
  nodes:
    - name: connected-endpoints-monitor-session-connected-endpoint-network-port-acl

servers:
  - name: INDIVIDUAL_1
    adapters:
      - switches: [connected-endpoints-monitor-session-connected-endpoint-network-port-acl]
        switch_ports: [Ethernet14]
        description: Monitor Ethernet 14
        monitor_sessions:
          - name: DMF
            role: source
            session_settings:
              access_group:
                type: ip
                name: acl1

network_ports:
  - switches:
      - connected-endpoints-monitor-session-connected-endpoint-network-port-acl
    switch_ports:
      - Ethernet14
    description: PC
    monitor_sessions:
      - name: DMF
        role: source
        source_settings:
          access_group:
            type: ip
            name: acl2

expected_error_message: >-
  Cannot set an ACL for both `session_settings` and `source_settings` under the monitor session 'DMF'
  for network_ports[0].

```

```bash
TASK [arista.avd.eos_designs : Generate device configuration in structured format] ***
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: pyavd._errors.AristaAvdInvalidInputsError: Cannot set an ACL for both `session_settings` and `source_settings` under the monitor session 'DMF' for network_ports[0].
fatal: [connected-endpoints-monitor-session-connected-endpoint-network-port-acl -> localhost]: FAILED! => {"changed": false, "msg": "Cannot set an ACL for both `session_settings` and `source_settings` under the monitor session 'DMF' for network_ports[0]."}

TASK [Assert eos_designs failed with the expected error message] ***************
ok: [connected-endpoints-monitor-session-connected-endpoint-network-port-acl] => {
    "changed": false,
    "msg": "All assertions passed"
}
```

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-sessions.md (hunk 9 above)
Comment (laxmikantchintakindi) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-sessions.md:97]: Yes, issue is on line 71. Shall I print this from 68 to 71. As ip access group ipv4ACL is present on myMonitoringSession2.

| Ethernet3, Ethernet5 | rx | ip | ipv4ACL | - |
| Ethernet10-15 | rx | ip | ipv4ACL | - |
| Ethernet12 | rx | ip | ipv4ACL | - |
| Ethernet18 | tx | ip | ipv4ACL | 100 |

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-sessions.md:97]: No you should just print the - if we are ignoring the ACL like we do in the EOS config.

**Comment by ClausHolbechArista** [comment]
> Would it be useful to have a 3rd negative test case where part of the config is coming from `connected_endpoints` and other part is coming from `network_ports`?

To not add too many tests for corner cases, this scenario could be the only one, since it would work the same no matter where the config is coming from.

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-sessions.md (hunk 9 above)
Comment (laxmikantchintakindi) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/monitor-sessions.md:97]: Fixed.

**Review by ClausHolbechArista — APPROVED** [review]

**Comment by ClausHolbechArista** [comment]
@alexeygorbunov remember to use `request changes` on the reviews if you want something changed, since it will block merging the PR. I forgot about the updated test, but it can go in a separate "Test" PR.
